compaction/compaction.cc

```diff
@@ -1898,4 +1898,18 @@ uint64_t compaction_descriptor::sstables_size() const {
     return boost::accumulate(sstables | boost::adaptors::transformed(std::mem_fn(&sstables::sstable::data_size)), uint64_t(0));
 }
 
+void compaction_descriptor::retrieve_owned_ranges_if_required(::compaction::compaction_state& cs) {
+    requires_cleanup = false;
```

Comment:
@bhalevy I think we can move this function into a static one living in compaction_manager.cc, rename it into:

static bool check_sstables_need_cleanup(const std::vector<sstables::shared_sstable>& sstables, ::compaction::compaction_state& cs);

Then return value can be assigned to compaction_descriptor::requires_cleanup.

Reply:
If we moved compaction_state out of compaction_manager for using in compaction_descriptor, I suggest we move it back, and make check_sstables_need_cleanup() (suggested above) a private static member of compaction_manager, which has access to compaction_state. compaction_state is not very meaninful outside compaction_manager, we even need a better name :-)

Reply:
Misread the function a little bit. It seems to me that having both requires_cleanup and owned_ranges is redundant. 

Regular compaction will not have owned_ranges unless explicitly set.

So why not just set owned_ranges if we want cleanup to be performed? Then need for cleanup is true based on the presence of owned_ranges. compaction_descriptor can have a new method needs_cleanup() which retrue if owned_ranges is set.

Reply:
retrieve_owned_ranges_if_required() can still be moved to compaction_manager, and its signature changed to:

owned_ranges_ptr compaction_manager::retrieve_owned_ranges_..(const compaction_descriptor& cd, compaction_state& cs);

Returns nullptr if descriptor doesn't need cleanup.

By the way, compaction_manager::task::compact_sstables() can avoid calling it if descriptor already has it set.

Reply:
Hmm, compaction_state is indirectly coupled with compaction_groups (via cg.as_table_state)
and the sstables that require cleanup "live" in the scope of the compaction_group and will be managed usually by compaction tasks that operate on that table_state.

The truncate case you mentioned in a review comment is special, but it can and probably should iterate through all compaction groups in the table to update them with the deleted sstables.

In the farther future, the plan is to keep a "truncate" or token-range tombstone that will logically delete everything at the given token range before the specified timestamp and truncate would just create those tombstones and kick-start cleanup (or major for that matter) compaction to get rid of the mass-deleted data.
Most sstables will be fully-expired and therefore immediately deleted - going through the normal compaction cleanup path that will remove them from the "sstables_requires_cleanup" set.  If memtable flush happened during truncate, the flushed sstable may contain data before and after the token-range tombstone and will need to go through normal compaction to keep only the valid data.

The same goes for tablet splitting which can use such a token-range tombstone for cleanup post tablet split.
/Cc @tgrabiec 

Bottom line, is that making it global is a bad idea IMO... :)

Reply:
Actually, compaction_state is compaction manager's internal per table-state metadata. 

The flow is the following:

1) compaction manager is called to run a job

2) it builds a compaction_descriptor, describing how compaction must be performed. Please note compaction_descriptor is only a descriptor, nothing else.

3) compaction is called with compaction_descriptor.

compaction_state is even a terrible name by the way. I have been thinking about renaming it recently.

When you pass compaction_state to compaction_descriptor, we're adding a cyclic dependency between compaction_descriptor and compaction manager's internal state. 

compaction_state is compaction manager's internal state only, doesn't make sense to expose it.

It makes much more sense for compaction_state (ouch, it has such a terrible name) to implement method retrieve_owned_ranges_if_required() and receive compaction_descriptor as parameter. After all, it's compaction manager that manages (pun intended) the set of SSTables requiring cleanup. Again, that's internal state and we really don't want to expose it. It's a layer violation.

Reply:
logic moved to compaction_manager::cleanup_sstables in v4